docs/develop/python/message-passing.mdx

```diff
@@ -27,155 +27,171 @@ tags:
   - dynamic-handlers
 ---
 
-This page shows how to do the following:
+See [Workflow message passing](/encyclopedia/workflow-message-passing) for a non–Python-specific
+introduction to using messages with Temporal Workflows.
 
-- [Develop with Signals](#signals)
-- [Develop with Queries](#queries)
-- [Develop with Updates](#updates)
-- [Dynamic Handler](#dynamic-handler)
+## Writing message handlers
 
-## Signals {#signals}
+The code samples below form part of a complete working sample, viewable at [message_passing/introduction](https://github.com/temporalio/samples-python/tree/message-passing/message_passing/introduction).
 
-**How to develop with Signals using the Python SDK.**
+General guidelines for writing message handlers:
 
-A [Signal](/encyclopedia/workflow-message-passing#sending-signals) is a message sent asynchronously to a running Workflow Execution which can be used to change the state and control the flow of a Workflow Execution.
-It can only deliver data to a Workflow Execution that has not already closed.
-
-Signals are defined in your code and handled in your Workflow Definition.
-Signals can be sent to Workflow Executions from a Temporal Client or from another Workflow Execution.
-
-There are two steps for adding support for a Signal to your Workflow code:
+- Arguments and return values of handlers (and the main Workflow function) must be [serializable](/dataconversion): a [dataclass](https://docs.python.org/3/library/dataclasses.html) will often be the right choice.
 
-1. **[Defining the Signal](#define-signal)** - You specify the name and data structure used by Temporal Clients when sending the Signal.
-2. **[Handling the Signal](#handle-signal)** - You write code that will be invoked when the Signal is received from a Temporal Client.
+- It's possible to write handler methods that take multiple arguments, but this is not recommended: instead use a single dataclass argument to which fields can be added as needed.
```

Comment:
May want to also mention that you can include your handlers anywhere in your Workflow definition since this isn't true of all SDKs.

Reply:
Can you expand on that @axfelix? (In Python handlers must be methods on the Workflow class annotated with one of the decorators.)

Reply:
Yes -- they have to be methods on the Workflow class, but they can be defined anywhere. Compared to eg Go, where iirc there's no such thing as a handler other than inline.

Reply:
I've added this bullet right at the top:

```markdown
General guidelines:

- Message handlers are defined as methods on the Workflow class, using one of the three decorators: `@workflow.query`, `@workflow.signal`, and `@workflow.update`.
```

Does that cover it? I think once Python programmers understand that they are methods, we can leave it to them to think of the possibilities for how they can define them. I.e. ranging from beginners who'll only know how to do it exactly as in our examples, to people who might choose inheritance, or mixins, or to inject them via class decorators or metaclasses!